staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -343,18 +343,24 @@ func (w *watchCache) updateCache(event *watchCacheEvent) {
 }
 
 // resizeCacheLocked resizes the cache if necessary:
-// - increases capacity by 2x if cache is full and all cached events occurred within last eventFreshDuration.
-// - decreases capacity by 2x when recent quarter of events occurred outside of eventFreshDuration(protect watchCache from flapping).
+// - increases capacity by [ eventFreshDuration / (eventTime - 1st event's timestamp) ]
+// 	 if cache is full and all cached events occurred within last eventFreshDuration.
+// - decreases capacity by [ eventFreshDuration / (eventTime - recent quarter event's timestamp) ] / 2
+//   when recent quarter of events occurred outside of eventFreshDuration(protect watchCache from flapping).
 func (w *watchCache) resizeCacheLocked(eventTime time.Time) {
 	if w.isCacheFullLocked() && eventTime.Sub(w.cache[w.startIndex%w.capacity].RecordTime) < eventFreshDuration {
-		capacity := min(w.capacity*2, w.upperBoundCapacity)
+		timeDiff := float64(eventFreshDuration) / float64(eventTime.Sub(w.cache[w.startIndex%w.capacity].RecordTime))
+		increase := max(int(timeDiff), 2)
+		capacity := min(w.capacity*increase, w.upperBoundCapacity)
```

Comment:
I'm sorry - I lost track of this PR.

I think this is a reasonable improvement. But this change requires unit tests to be mergable.